### PR TITLE
Fix: Fix incorrect casing when rendering fields and enums 

### PR DESCRIPTION
**Pull request by sanjayjayaramu-cruise**
Fixes https://github.com/marcosschroh/dataclasses-avroschema/issues/693 https://github.com/marcosschroh/dataclasses-avroschema/issues/692
This change fixed incorrect case rendering of fields inside a record which has the same record type and uses a reference of the type to determine the type of the field. Since the field name is converted to Pascal case for Record types, the reference type name should also be Pascal cased.
The Enum class rendering has a similar issue where the default value of the enum field does not contain the Pascal case of the Enum class.

### CODE REVIEW

**Comment by sanjayjayaramu-cruise** [comment]
> @sanjayjayaramu-cruise Could you fix the tests?

Still trying to get tests work locally, will fix them asap

**Comment by sanjayjayaramu-cruise** [comment]
> @sanjayjayaramu-cruise Could you fix the tests?

Hey @marcosschroh , the problem with my approach here is that the round trip tests now fail generate the original schema since the change converts basically all dataclasses and enums to use pascal case. May be this is not desired and works for everyone?

**Comment by sanjayjayaramu-cruise** [comment]
> > @sanjayjayaramu-cruise Could you fix the tests?
> 
> Hey @marcosschroh , the problem with my approach here is that the round trip tests now fail tp generate the original schema from model classes since the change converts basically all dataclasses and enums to use pascal case. May be this is not desired and works for everyone?

Also note that, the round trip test(s) will fail for any nested record type on master and there is not unit test covering that. I think there is no 100% guarantee that the user can get back the schema using the model generator with the same casing as the original schema.  
Here is an example schema and the resulting schema generated from the autogen class

```
    original_schema = {
        'type': 'record', 
        'name': 'prism', 
        'fields': [
            {
             'name': 'pose', 
             'type': {
                 'type': 'record', 
                 'name': 'pose', 
                 'fields': [{
                     'name': 'pose_name', 
                     'type': 'string'
                     }]
                }
            }, 
        ]
    }
```
After
```
{
    "type": "record",
    "name": "Prism",
    "fields": [
        {
            "name": "pose",
            "type": {
                "type": "record",
                "name": "Pose",
                "fields": [
                    {
                        "name": "pose_name",
                        "type": "string"
                    }
                ]
            }
        }
    ]
```
}

**Comment by marcosschroh** [comment]
Hey @sanjayjayaramu-cruise 

The problem that you are facing is `overshadowed`. It will never work because in python you can not have this:

```python
@dataclasses.dataclass
class MslPrism(AvroModel):
    """
    The 3D bounding box labels of this object.
    """
    data_source: data_source = dataclasses.field(metadata={'doc': 'Source of the ingested segment, ie. ON_ROAD, SIMULATION, etc.'}, default=data_source.UNKNOWN)
```

The problem is described in the [docs](https://marcosschroh.github.io/dataclasses-avroschema/model_generator/#malformed-schemas) and it is independent of this library, the same happens in `pydantic` or any library. The field name CAN NOT be the same as the type.

There are 2 ways of mitigate this:

1. Have a "proper schema", making sure that the `field name` is different from the `type`
2. Modify your PR and add another property the `enum` call `schema_name` to override the `class.__name__`.

By the way, there are tests for record relationship: [OneToOne](https://github.com/marcosschroh/dataclasses-avroschema/blob/master/tests/model_generator/test_model_generator.py#L411C5-L411C40), [OneToMany](https://github.com/marcosschroh/dataclasses-avroschema/blob/master/tests/model_generator/test_model_generator.py#L443) using `array`, [OneToMany](https://github.com/marcosschroh/dataclasses-avroschema/blob/master/tests/model_generator/test_model_generator.py#L474) using `maps` and [Self Relatinship](https://github.com/marcosschroh/dataclasses-avroschema/blob/master/tests/model_generator/test_model_generator.py#L474)

**Comment by sanjayjayaramu-cruise** [comment]
> Modify your PR and add an extra property the enum generated, called schema_name. This new property can be used to generate the enum name rather than class.__name__. Similar to what is describe in the [docs](https://marcosschroh.github.io/dataclasses-avroschema/records/#class-meta) for Records and Meta.schema_name
Could you expand a little on this? Do you mean, only when the field name == type name, add a new Meta class to retain the schema main when we deserialize it for the tests?

**Comment by marcosschroh** [comment]
> Could you expand a little on this? Do you mean, only when the field name == type name, add a new Meta class to retain the schema main when we deserialize it for the tests?

Yes and when the `enum name` does not follow the `class convention` naming in python. We should do the same for `records`.

-  If an `enum name` of `record name` defined in a `avro schema` does not follow the class naming convention in python, then we should add the property `schema_name` to `class Meta` which must contain the original name. If we do this then we can go from `schema <--> Model`.
- If an `enum name` is the same as the `field name`, then we do the same. Possible for records as well. 

There might be cases where the `field_name` and `inner name` (record name` of `enum name`) follow the `python naming` convention and they are the both equal. For this use cases the end user must solve it, either modifying the schema are using the property `schema_name`.

**Comment by marcosschroh** [comment]
> Could you expand a little on this? Do you mean, only when the field name == type name, add a new Meta class to retain the schema main when we deserialize it for the tests?

Yes and when the `enum name` does not follow the `class convention` naming in python. We should do the same for `records`.

-  If an `enum name` or `record name` defined in a `avro schema` does not follow the class naming convention in python, then we should add the property `schema_name` to `class Meta` which must contain the original name. If we do this then we can go from `schema <--> Model`.
- If an `enum name` is the same as the `field name`, then we do the same. Possible for records as well. 

There might be cases where the `field_name` and `inner name` (`record name` or `enum name`) follow the `python naming` convention and they are the both equal. For this use cases the end user must solve it, either modifying the schema are using the property `schema_name`.

**Comment by sanjayjayaramu-cruise** [comment]
Hey @marcosschroh I have updated the PR with the suggestions. PTAL when you can. Thank you.

File: dataclasses_avroschema/fields/fields.py [hunk 1]
```diff
@@ -488,6 +488,10 @@ def validate_symbols(self) -> bool:
     def get_avro_type(self) -> typing.Union[str, types.JsonDict]:
         metadata = self._get_meta_class_attributes()
         name = self.type.__name__
+        # If the enum name in avro schema different than the python enum class name, use the schema_name.
+        if "schema_name" in metadata:
+            name = metadata.get("schema_name")
+            metadata.pop("schema_name")
```
Comment (marcosschroh) [inline dataclasses_avroschema/fields/fields.py:494]: You can do `name = metadata.pop("schema_name", self.type.__name__)`

File: dataclasses_avroschema/model_generator/lang/python/base.py [hunk 2]
```diff
@@ -448,7 +460,12 @@ def parse_enum(self, field: JsonDict) -> str:
 
         docstring = self.render_docstring(docstring=field.get("doc"))
         enum_class = templates.enum_template.safe_substitute(name=enum_name, symbols=symbols_repr, docstring=docstring)
-        metaclass = self.render_metaclass(schema=field, decorator=templates.METACLASS_DECORATOR)
+        add_schema_name = not enum_name == field["name"]
```
Comment (marcosschroh) [inline dataclasses_avroschema/model_generator/lang/python/base.py:463]: You can do: `add_schema_name = enum_name != field["name"]`

File: dataclasses_avroschema/model_generator/lang/python/base.py [hunk 3]
```diff
@@ -185,7 +196,8 @@ def render_class(self, *, schema: JsonDict) -> str:
             docstring=docstring,
         )
 
-        class_metadata = self.render_metaclass(schema=schema, field_order=field_order)
+        add_schema_name = not name == schema["name"]
```
Comment (marcosschroh) [inline dataclasses_avroschema/model_generator/lang/python/base.py:199]: You can do: `add_schema_name = name != schema["name"]`

File: tests/model_generator/test_model_pydantic_generator.py [hunk 4]
```diff
@@ -185,14 +186,18 @@ class Meta:
...
+    expected_result = f"""
 from dataclasses_avroschema import types
 from dataclasses_avroschema.pydantic import AvroBaseModel
 
 
 
 class Demo(AvroBaseModel):
     foo: types.condecimal(max_digits=10, decimal_places=3)
+
+    {templates.METACLASS_DECORATOR}
```
Comment (marcosschroh) [inline tests/model_generator/test_model_pydantic_generator.py:198]: `{templates.METACLASS_DECORATOR}` is not needed. It is used only when `enums` has a `class Meta` (depending on the python version)

**Review by marcosschroh — COMMENTED** [review]
It looks good @sanjayjayaramu-cruise . Some small changes and we are good to go.

File: tests/model_generator/test_model_pydantic_generator.py (hunk 4 above)
Comment (sanjayjayaramu-cruise) [reply tests/model_generator/test_model_pydantic_generator.py]: Agreed!

File: dataclasses_avroschema/model_generator/lang/python/base.py (hunk 3 above)
Comment (sanjayjayaramu-cruise) [reply dataclasses_avroschema/model_generator/lang/python/base.py]: Agreed!

File: dataclasses_avroschema/model_generator/lang/python/base.py (hunk 2 above)
Comment (sanjayjayaramu-cruise) [reply dataclasses_avroschema/model_generator/lang/python/base.py]: Agreed!

File: dataclasses_avroschema/fields/fields.py (hunk 1 above)
Comment (sanjayjayaramu-cruise) [reply dataclasses_avroschema/fields/fields.py]: Nice. Agreed!

**Comment by sanjayjayaramu-cruise** [comment]
> It looks good @sanjayjayaramu-cruise . Some small changes and we are good to go.

Thank you. Applied the recommended changes.

**Review by marcosschroh — APPROVED** [review]

**Pull request merged**